modin/data_management/functions/groupby_function.py

```diff
@@ -437,6 +534,123 @@ def wrapper(df):
+            Name of the groupby function. This is a hint to be able to do special casing.
+            Note: this parameter is a legacy from the ``groupby_size`` implementation,
+            it's a hacky one and probably will be removed in the future.
+
+        Returns
+        -------
+        drop : bool
+            Indicates whether to drop all index levels (True) or insert them into the
+            resulting columns (False).
+        lvls_to_drop : list of ints
+            Contains numeric indices of the levels of the result index to drop as intersected.
+
+        Examples
+        --------
+        >>> groupby_result = compute_groupby_without_processing_as_index_parameter()
+        >>> if not as_index:
+        >>>     drop, lvls_to_drop = handle_as_index(**extract_required_params(groupby_result))
+        >>>     if len(lvls_to_drop) > 0:
+        >>>         groupby_result.index = groupby_result.index.droplevel(lvls_to_drop)
+        >>>     groupby_result_with_processed_as_index_parameter = groupby_result.reset_index(drop=drop)
+        >>> else:
+        >>>     groupby_result_with_processed_as_index_parameter = groupby_result
+        """
+        if partition_idx != 0 or (not drop and method != "size"):
+            return True, []
+
+        if method == "size":
+            return False, []
+
+        if selection is not None:
+            selection = selection if is_list_like(selection) else (selection,)
+            if partition_selection is None:
+                partition_selection = selection
+            partition_selection = (
+                partition_selection
+                if is_list_like(partition_selection)
+                else (partition_selection,)
+            )
+            if len(result_cols) != len(partition_selection):
+                cols_failed_to_select = pandas.Index(partition_selection).difference(
+                    result_cols
+                )
+                # If some of the selected 'by' columns were dropped as non-suitable for
+                # the current aggregation it's allowed to insert them back. Dropping
+                # them from the selection to be able to do so:
+                selection = pandas.Index(selection).difference(cols_failed_to_select)
+
+        if not isinstance(internal_by_cols, pandas.Index):
+            if not is_list_like(internal_by_cols):
+                internal_by_cols = [internal_by_cols]
+            internal_by_cols = pandas.Index(internal_by_cols)
+
+        internal_by_cols = (
+            internal_by_cols[~internal_by_cols.str.match(r"__reduced__.*", na=False)]
+            if hasattr(internal_by_cols, "str")
+            else internal_by_cols
+        )
+
+        # We want to insert such internal-by-cols that are not presented
+        # in the result in order to not create naming conflicts
+        cols_to_insert = (
+            internal_by_cols.copy()
+            if selection is None and internal_by_cols.nlevels != result_cols.nlevels
```

Comment:
if `internal_by_cols.nlevels != result_cols.nlevels` we don't need to compute the difference, there won't be any intersections anyway